src/coreclr/jit/morph.cpp

```diff
@@ -10242,13 +10242,18 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 if (op1->OperIs(GT_NEG) && !op1->gtGetOp1()->IsCnsIntOrI() && op2->IsCnsIntOrI() &&
                     !op2->IsIconHandle())
                 {
-                    GenTree* newOp1   = op1->gtGetOp1();
-                    GenTree* newConst = gtNewIconNode(-op2->AsIntCon()->IconValue(), op2->TypeGet());
-                    DEBUG_DESTROY_NODE(op1);
-                    DEBUG_DESTROY_NODE(op2);
-                    tree->AsOp()->gtOp1 = newOp1;
-                    tree->AsOp()->gtOp2 = newConst;
-                    return fgMorphSmpOp(tree, mac);
+                    // We are changing the shape of a child node here which
+                    // means we need to recall fgMorphTree (in particular to
+                    // do promotion-specific morphing of op1); however, morphing
+                    // the top-level node twice is not ok. Reuse op1 as the new
+                    // top-level node instead.
+                    op2->AsIntCon()->gtIconVal = -op2->AsIntCon()->gtIconVal;
+                    op1->ChangeOper(GT_MUL);
+                    op1->SetVNsFromNode(tree);
+                    op1->AsOp()->gtOp2 = op2;
+                    DEBUG_DESTROY_NODE(tree);
+
+                    return fgMorphTree(op1);
```

Comment:
Another option here would be to not return early and carry on morphing instead (would require updating `op1` and `op2` local variables).

Reply:
I don't think that would fix the problem. As I tried to explain in the comment we morph promoted/demoted locals only when morphing their user.

Reply:
You are right, sorry for the distraction.

Coincidentally, after #69819 and the equivalent change for implicit byrefs, that will no longer be a problem.

Reply:
> Coincidentally, after https://github.com/dotnet/runtime/pull/69819 and the equivalent change for implicit byrefs, that will no longer be a problem.

Regardless, it feels iffy to leave the call to `fgMorphSmpOp` which somehow implies that it is fully sufficient as a replacement of `fgMorphTree`. I guess there are other cases where `fgMorphTree` will make additional decisions depending on child nodes, e.g. local assertion prop.

Reply:
Yes, I agree with that.

(The usage of `fgMorphSmpOp` instead of `fgMorphTree` in these code paths is a workaround for the double-morphing assert)